### PR TITLE
skip not ready nodes in networking test

### LINKED ISSUE #9345: e2e networking test schedule a pod to a NotReady node
I ran e2e tests against docker 1.7.0-rc2 release, found a bunch of issues. Here is one I think it belongs to kubernetes for sure. 

1) ResizeNode test failed, the machine is removed from instance group:

```
| e2e-test-dawnchen-master      | us-central1-b | RUNNING | 10.240.156.229 | 104.154.64.108  |
+-------------------------------+---------------+---------+----------------+-----------------+
| e2e-test-dawnchen-minion-vxvg | us-central1-b | RUNNING | 10.240.146.175 | 104.197.40.212  |
+-------------------------------+---------------+---------+----------------+-----------------+
| e2e-test-dawnchen-minion-wkpa | us-central1-b | RUNNING | 10.240.211.202 | 104.197.91.127  |
```

But kubectl get nodes still returns 3 ndoes:
```
$ kubectl get nodes
NAME                            LABELS                                                 STATUS
e2e-test-dawnchen-minion-5l11   kubernetes.io/hostname=e2e-test-dawnchen-minion-5l11   NotReady
e2e-test-dawnchen-minion-vxvg   kubernetes.io/hostname=e2e-test-dawnchen-minion-vxvg   Ready
e2e-test-dawnchen-minion-wkpa   kubernetes.io/hostname=e2e-test-dawnchen-minion-wkpa   Ready
```
This might be caused by the new docker binary I am testing against here. This is not the issue I am filing here.

2) Now networking test is running later, I noticed that the test is putting a pod on each node:  nettest-gwtc4 on node e2e-test-dawnchen-minion-5l11 

```
Networking 
  should function for intra-pod communication
  /go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/networking.go:203
[BeforeEach] Networking
  /go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/framework.go:47
STEP: Creating a kubernetes client
>>> testContext.KubeConfig: /usr/local/google/home/dawnchen/.kube/config
STEP: Building a namespace api object
STEP: Waiting for a default service account to be provisioned in namespace
INFO: Waiting up to 2m0s for service account default to be provisioned in ns e2e-tests-nettest-2a0e3f3f-102f-4ca9-b8a0-118d9da29626
INFO: Service account default in ns e2e-tests-nettest-2a0e3f3f-102f-4ca9-b8a0-118d9da29626 found. (39.623658ms)
[BeforeEach] Networking
  /go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/networking.go:51
STEP: Executing a successful http request from the external internet
[It] should function for intra-pod communication
  /go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/networking.go:203
STEP: Creating a service named "nettest" in namespace "e2e-tests-nettest-2a0e3f3f-102f-4ca9-b8a0-118d9da29626"
STEP: Creating a webserver (pending) pod on each node
INFO: Created pod nettest-gwtc4 on node e2e-test-dawnchen-minion-5l11
INFO: Created pod nettest-0awly on node e2e-test-dawnchen-minion-vxvg
INFO: Created pod nettest-0qbe9 on node e2e-test-dawnchen-minion-wkpa
STEP: Waiting for the webserver pods to transition to Running state
INFO: Waiting up to 5m0s for pod nettest-gwtc4 status to be running
INFO: Waiting for pod 'nettest-gwtc4' in namespace 'e2e-tests-nettest-2a0e3f3f-102f-4ca9-b8a0-118d9da29626' status to be '"running"' (found phase: '"Pending"', readiness: false) (41.642383ms)
INFO: Waiting for pod 'nettest-gwtc4' in namespace 'e2e-tests-nettest-2a0e3f3f-102f-4ca9-b8a0-118d9da29626' status to be '"running"' (found phase: '"Pending"', readiness: false) (5.083175533s)
INFO: Waiting for pod 'nettest-gwtc4' in namespace 'e2e-tests-nettest-2a0e3f3f-102f-4ca9-b8a0-118d9da29626' status to be '"running"' (found phase: '"Pending"', readiness: false) (10.125042821s)
INFO: Waiting for pod 'nettest-gwtc4' in namespace 'e2e-tests-nettest-2a0e3f3f-102f-4ca9-b8a0-118d9da29626' status to be '"running"' (found phase: '"Pending"', readiness: false) (15.166074216s)
INFO: Waiting for pod 'nettest-gwtc4' in namespace 'e2e-tests-nettest-2a0e3f3f-102f-4ca9-b8a0-118d9da29626' status to be '"running"' (found phase: '"Pending"', readiness: false) (20.207430126s)
INFO: Waiting for pod 'nettest-gwtc4' in namespace 'e2e-tests-nettest-2a0e3f3f-102f-4ca9-b8a0-118d9da29626' status to be '"running"' (found phase: '"Pending"', readiness: false) (25.248069231s)
INFO: Waiting for pod 'nettest-gwtc4' in namespace 'e2e-tests-nettest-2a0e3f3f-102f-4ca9-b8a0-118d9da29626' status to be '"running"' (found phase: '"Pending"', readiness: false) (30.289449349s)
INFO: Waiting for pod 'nettest-gwtc4' in namespace 'e2e-tests-nettest-2a0e3f3f-102f-4ca9-b8a0-118d9da29626' status to be '"running"' (found phase: '"Pending"', readiness: false) (35.330526598s)

```

I haven't checked networking code in detail yet, not sure the issue is test itself, or scheduler issue. cc/ @lavalamp since you recently changed this. 

cc/ @quinton-hoole

### PR DESCRIPTION
**Pull request by marekbiskup**
fixes #9345

### CODE REVIEW

**Comment by dchen1107** [comment]
LGTM

File: test/e2e/util.go [hunk 1]
```diff
@@ -1144,6 +1144,21 @@ func waitForNodeToBeNotReady(c *client.Client, name string, timeout time.Duratio
 	return waitForNodeToBe(c, name, false, timeout)
 }
 
+func isNodeReadySetAsExpected(node *api.Node, wantReady bool) bool {
```
Comment (lavalamp) [inline test/e2e/util.go]: suggest just "isNodeReady" and omit the wantReady parameter. YAGNI

File: test/e2e/networking.go [hunk 2]
```diff
@@ -116,6 +116,14 @@ var _ = Describe("Networking", func() {
 		if err != nil {
 			Failf("Failed to list nodes: %v", err)
 		}
+		// previous tests may have cause failures of some nodes. Let's skip
+		// 'Not Ready' nodes, just in case (there is no need to fail the test).
+		filterNodes(nodes, func(node api.Node) bool {
+			return isNodeReadySetAsExpected(&node, true)
+		})
+		if len(nodes.Items) == 0 {
```
Comment (lavalamp) [inline test/e2e/networking.go]: < 2

This test can't work if there's fewer than two nodes.

File: test/e2e/util.go (hunk 1 above)
Comment (lavalamp) [inline test/e2e/util.go]: nvm I see it's used... sigh

File: test/e2e/util.go [hunk 3]
```diff
@@ -1157,22 +1172,27 @@ func waitForNodeToBe(c *client.Client, name string, wantReady bool, timeout time
...
+			return true
 		}
 	}
 	Logf("Node %s didn't reach desired readiness (%t) within %v", name, wantReady, timeout)
 	return false
 }
 
+// Filters nodes in NodeList in place, removing nodes that do not
+// satisfy the given condition
+func filterNodes(nodeList *api.NodeList, fn func(node api.Node) bool) {
```
Comment (lavalamp) [inline test/e2e/util.go]: Can you add a TODO: consider merging with pkg/client/cache.NodeLister

https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/client/cache/listers.go#L111

File: test/e2e/networking.go (hunk 2 above)
Comment (dchen1107) [inline test/e2e/networking.go]: Good catch!

**Comment by marekbiskup** [comment]
@quinton-hoole, Is Jenkins stable? Shall I take the results seriously?

**Comment by lavalamp** [comment]
LGTM - I think the failures are not related to this PR.

**Comment by quinton-hoole** [comment]
Agreed, LGTM

**Comment by satnam6502** [comment]
Please can this be merged soon-ish since I am blocked on it for fixing the Elasticsearch logging test? Thanks.

**Comment by marekbiskup** [comment]
I suppose that, previously, the test did no checks with one node, and now it is failing. That's the reason for @jayunit100's objections.

@lavalamp, you said this test cannot work with 1 node. Maybe I can create two pods on the same node, if there is only one? Will this test also be reasonable (I understand it will be limited).

**Comment by lavalamp** [comment]
The whole point of this test is to verify that the network works between two pods on different machines. I can see making another version of the test that makes sure the network works on a single machine, but that would be a different test.

**Comment by jayunit100** [comment]
Since when? This test generally was used to confirm that all nodes could contact one another over the service endpoint... Iirc that is still a valid test on a single node, and indeed, by breaking ip tables or other underlying services you can toggle the passing of the test on or off , making it a great test for 1node dev clusters.

**Comment by lavalamp** [comment]
Since I wrote this test originally, I feel qualified to claim that the purpose of this test is to verify that packets correctly travel between nodes. :)

Please note that the test load reads the endpoints and tries to dial the other pods directly. https://github.com/GoogleCloudPlatform/kubernetes/blob/master/contrib/for-tests/network-tester/webserver.go#L218

The service is only used for discovery.

**Comment by lavalamp** [comment]
I do see that there may be value in a version of this test that runs on a single node. But in our continuous e2e runs, for GCE, GKE, and other cloud providers--we absolutely need to be failing if we can't get packets between two nodes.

**Comment by jayunit100** [comment]
touche` :) but  indeed, it also contacts itself through the service endpoint.  hence it may have been written for the purpose of 2 nodes, but its an awesome test even on a single node ! security, networking, and certificates are **all** tested by the single node networking test.  You're baby is more useful than you originally thought ! :)

**Comment by lavalamp** [comment]
Yes, it exercises a lot of useful stuff. I'm fine with, e.g., gating the two node requirement with a comparison to the cloud provider.

**Comment by jayunit100** [comment]
ok awesome .  that is a good way to do it.

**Comment by marekbiskup** [comment]
I created #10012 for the purpose of allowing the test to run on a (local) single node

**Pull request merged**